### PR TITLE
Memoize promises instead of results to avoid concurrency problems

### LINKED ISSUE #34: [Discussion] How to dedupe concurrent memoizable requests?
## The Problem
Concurrent requests usually happen:
- When resolving multiple fields of the same object.
- When one object appears in multiple locations of a query.

Multiple requests occur and these requests should be deduped if marked as memoizable.

## Examples
```typescript
const resolvers: Resolvers = {
  UserInfo: {
    name: async ({ mid }, _, { dataSources }) =>
      (await dataSources.example.getUserInfo(mid)).name,
    gender: async ({ mid }, _, { dataSources }) =>
      (await dataSources.example.getUserInfo(mid)).gender,
    avatar: async ({ mid }, _, { dataSources }) =>
      (await dataSources.example.getUserInfo(mid)).avatar,
  },
}
```
```graphql
query Users {
  currentUser {
    mid
    name
    avatar
  }
  userInfo(mid: 123) {
    name
    avatar
  }
}
```

## My suggestion
We should use a lazy loader (like [`graphql/dataloader`](https://github.com/graphql/dataloader) but much simpler as we don't need batching and caching).

See #29 for my rough workaround.

## Minor question
How to define `response.memoize` if the response is from deduped result?

### PR DESCRIPTION
**Pull request by llc1123**
Fixes #34

### CODE REVIEW

**Comment by llc1123** [comment]
Anything blocking this PR? @StarpTech

**Comment by StarpTech** [comment]
@llc1123 please fix the conflict. I'm going to review it in the next days.

**Comment by llc1123** [comment]
@StarpTech fixed.

**Comment by bradleymorris** [comment]
Really looking forward to this PR - the memoization shortcoming that this PR addresses has been a blocker for us.

File: src/http-data-source.ts [hunk 1]
```diff
@@ -399,63 +419,53 @@ export abstract class HTTPDataSource<TContext = any> extends DataSource {
...
-        }
+  protected async trace<TResult>(request: Request, fn: () => Promise<TResult>): Promise<TResult> {
+    if (process.env.NODE_ENV === 'development') {
+      const startTime = Date.now()
+      try {
+        return await fn()
+      } finally {
+        const duration = Date.now() - startTime
+        const label = `${request.method || 'GET'} ${request.path}`
+        console.log(`${label} (${duration}ms)`)
```
Comment (StarpTech) [inline src/http-data-source.ts:465]: Please use the optional apollo logger.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -14,10 +14,13 @@ jobs:
     - uses: actions/setup-node@v1
       with:
         node-version: ${{ matrix.nodejs }}
+        
+    - name: Update npm
+      run: npm install -g npm@latest
```
Comment (tdipadova3rd) [inline .github/workflows/ci.yml:19]: why are you updating npm here? shouldn't the npm version be configured via the `setup-node` action?

it looks like this job uses the matrix strategy to test multiple node versions, and doesn't cache `npm`, so presumably we want to test against each version of `npm` shipped with the corresponding node version we test against?

Comment (llc1123) [reply .github/workflows/ci.yml:19]: The npm version packed with node 12 and 14 is too low to support lockfile v2

**Comment by llc1123** [comment]
Updated. @StarpTech

**Comment by tdipadova3rd** [comment]
Any updates @StarpTech ? Would love to use these changes.

**Comment by 0xTea** [comment]
> 



> Any updates @StarpTech ? Would love to use these changes.

same , waiting on this update 👍

File: src/http-data-source.ts [hunk 3]
```diff
@@ -399,63 +418,53 @@ export abstract class HTTPDataSource<TContext = any> extends DataSource {
   }
 
   private async request<TResult = unknown>(request: Request): Promise<Response<TResult>> {
-    if (Object.keys(request.query).length > 0) {
-      request.path = request.path + '?' + this.buildQueryString(request.query)
+    const options: Request = {
+      ...this.globalRequestOptions,
```
Comment (javi11) [inline src/http-data-source.ts:422]: Not sure if this is the intended behavior, but as was working before globalOptions override local request options not the other way around :S.

**Comment by StarpTech** [comment]
@tdipadova3rd sorry for the delay. I'll review and merge this PR this week :crossed_fingers:

File: .github/workflows/ci.yml (hunk 2 above)
Comment (StarpTech) [inline .github/workflows/ci.yml:19]: Why is this necessary?

File: benchmarks/README.md [hunk 4]
```diff
@@ -6,19 +6,19 @@ node http.js
...
+Results for 1000 subsequent requests:
+apollo-datasource-rest (http1) | total time: 1033621609ns (1033.622ms)
+apollo-datasource-http (http1) | total time: 589782921ns (589.783ms)
 ---
-apollo-datasource-http (http1) <> apollo-datasource-rest (http1) percent change: -56.277%
+apollo-datasource-http (http1) <> apollo-datasource-rest (http1) percent change: -42.940%
 ```
 
-**Result:** `apollo-datasource-http` is around `56%` faster than `apollo-datasource-rest`
+**Result:** `apollo-datasource-http` is around `43%` faster than `apollo-datasource-rest`
```
Comment (StarpTech) [inline benchmarks/README.md:24]: This would mean a performance drop of more than 10%.

Comment (StarpTech) [reply benchmarks/README.md:24]: I could verify it on my machine with even more (-15%). Can we improve this?

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by StarpTech** [comment]
@llc1123 I appreciate the refactoring but you touched way too many places which aren't related to the issue. This makes reviewing hard.

File: benchmarks/README.md (hunk 4 above)
Comment (llc1123) [reply benchmarks/README.md:24]: I think it's mainly because of the overhead of caching the promises and I don't think it is a problem to be solved.

**Comment by llc1123** [comment]
The changes are in separate commits and should be easy to be figured out.

File: benchmarks/README.md (hunk 4 above)
Comment (llc1123) [reply benchmarks/README.md:24]: Other possibilities:
1. The newer version of undici has lower performance. 
2. The node version is different than the last time you run the benchmark which improved the baseline.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (llc1123) [reply .github/workflows/ci.yml:19]: I've upgraded the dependencies and the lockfile is now in v2 (since npm@7), which npm included with old node environments don't support.

File: benchmarks/README.md (hunk 4 above)
Comment (StarpTech) [reply benchmarks/README.md:24]: 1. I verified the new undici version on main, and it has no performance drop.
2. No, I tested both with the same.

Please make your tests.

**Comment by StarpTech** [comment]
> The changes are in separate commits and should be easy to be figured out.

That's not what I meant.

**Review by JustinTRoss — CHANGES_REQUESTED** [review]
You really went all out with the improvements 👏 . It seems like this could be split into 3-4 PRs to deliver clear units of change and make clear how each unit is being tested and what the effects are.

**Comment by StarpTech** [comment]
@llc1123 in order to close the issue. Please create another PR that contains only the fix for https://github.com/StarpTech/apollo-datasource-http/issues/34. The next step is to avoid a performance drop of ~15%. I'll help you with that.

**Comment by bradleymorris** [comment]
What's the status on this?  Still chomping at the bit to cut over to this data source...

**Comment by llc1123** [comment]
I have migrated to apollo server v4 and this code is no longer maintained.
